### PR TITLE
[ENH] Implemented new endpoint to return known node APIs

### LINKED ISSUE #26: New endpoint to return known node APIs
- [x] Update `NB_NODES` env variable to also define name of node
- [x] rename `NB_NODES` to `LOCAL_NB_NODES` to differentiate from the public list
- [x] Include node label (name) at the dataset level in the response
- [ ] update .env file name to fed.env

Decisions:
- every f-API talks only to n-APIs, not other f-APIs (at most one level of separation b/w query-tool and n-APIs)
- a new public endpoint provides a list of known public n-APIs
- on startup, f-API looks up this endpoint and adds all the known n-APIs to its own index of n-APIs
- f-APIs can have a local-only index of n-APIs, e.g. when running a private n-API. This is configured via the `NB_NODES` `.env` variable
- f-APIs join both lists (from endpoint + .env) and take the unique entries, no duplicates

our current federate.neurobagel.org API has a list of n-APIs hardcoded. This info will be lifted out into the service endpoint running nodes.neurobagel.org
-> try making this as dumb as possible, e.g. just from some static file like on GH

This new public list is going to include only those nodes that are
- meant to be seen by everyone
- publicly, i.e. anonymously reachable

### PR DESCRIPTION
**Pull request by rmanaem**
- Closes #26 

Changes proposed in this pull request:

- Modified the format of the `NB_NODES` as a string of strings to string of tuples 
- Renamed `NB_NODES` to `LOCAL_NB_NODES` 
- Refactored and renamed `parse_nodes_as_list` to `parse_nodes_as_dict`
- Updated tests
- Updated `get` function in `crud.py`
- Fixed the parsing/creation of the output in `get_terms` function of `crud.py`
- Implemented `CohortQueryResponse`
- Added `response_model` for `get_query` endpoint
- Implemented `nodes` router and an endpoint to return all known node APIs

## Checklist

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see https://neurobagel.org/contributing/pull_requests for more info)_
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass

For new features:
- [ ] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

### CODE REVIEW

File: app/api/crud.py [hunk 1]
```diff
@@ -85,16 +89,17 @@ async def get_terms(data_element_URI: str):
...
-        term
-        for list_of_terms in cross_node_results
-        for term in list_of_terms[data_element_URI]
-    )
-    return {data_element_URI: list(unique_terms)}
+    unique_terms_dict = {}
+
+    for list_of_terms in cross_node_results:
+        for term in list_of_terms[data_element_URI]:
+            unique_terms_dict[term["TermURL"]] = term
```
Comment (surchs) [inline app/api/crud.py:103]: this works, but should we not do
```suggestion
            unique_terms_dict[term["TermURL"]] = term["Label"]
```
or whatever the key for the string value is? in other words: what's the expected shape of the returned list?

File: app/api/crud.py [hunk 2]
```diff
@@ -60,9 +60,13 @@ async def get(
     if image_modal:
         params["image_modal"] = image_modal
 
-    for node_url in util.parse_nodes_as_list(util.NEUROBAGEL_NODES):
+    nodes_dict = util.parse_nodes_as_dict(util.NEUROBAGEL_NODES)
+    for node_url in nodes_dict.keys():
```
Comment (surchs) [inline app/api/crud.py:64]: you can use `.items()` here to extract keys and values transparently

File: app/api/utility.py [hunk 3]
```diff
@@ -1,23 +1,28 @@
...
+def parse_nodes_as_dict(nodes: str) -> list:
+    """Returns user-defined Neurobagel nodes as a dict.
+    It uses a regular expression to match the url, name pairs.
     Makes sure node URLs end with a slash."""
-    nodes_list = nodes.split(" ")
-    for i in range(len(nodes_list)):
-        if nodes_list[i] and not nodes_list[i].endswith("/"):
-            nodes_list[i] += "/"
-    return list(filter(None, nodes_list))
+    pattern = re.compile(r"\((https?://[^\s]+), ([^\)]+)\)")
```
Comment (surchs) [inline app/api/utility.py:19]: I think you can simplify this with named capture groups
```suggestion
    pattern = re.compile(r"(?P<url>https?://[^\s]+), (?P<label>[^\)]+)")
```

And then

```python
url = match.group('url')
label = match.group('label')
```

File: app/api/utility.py [hunk 4]
```diff
@@ -1,23 +1,28 @@
...
+    matches = pattern.findall(nodes)
+    for i in range(len(matches)):
+        if not matches[i][0].endswith("/"):
+            matches[i] = (matches[i][0] + "/", matches[i][1])
+    nodes_dict = {url: name for url, name in matches}
```
Comment (surchs) [inline app/api/utility.py:24]: see above for named capture groups.

File: tests/test_utility.py [hunk 5]
```diff
@@ -9,30 +9,35 @@
...
         (
-            " https://firstnode.neurobagel.org/query/ https://secondnode.neurobagel.org/query/  ",
-            [
-                "https://firstnode.neurobagel.org/query/",
-                "https://secondnode.neurobagel.org/query/",
-            ],
+            "(firstnode.neurobagel.org/query/, firstnode) (https://secondnode.neurobagel.org/query, secondnode)",
+            {
+                "https://secondnode.neurobagel.org/query/": "secondnode",
+            },
```
Comment (surchs) [inline tests/test_utility.py:26]: if we don't start with https we want to ignore? I think that should also give a warning in the API log when the .env is read - i.e.: we read some stuff from you that we can't understand, so we'll ignore

File: tests/test_utility.py [hunk 6]
```diff
@@ -9,30 +9,35 @@
...
+        ),
+        (
+            "( , firstnode) (https://secondnode.neurobagel.org/query, secondnode)",
```
Comment (surchs) [inline tests/test_utility.py:29]: I would split these cases off into a "bad input" type test and then both assert that the tool can process it (as now) but also that we raise a warning

**Review by surchs — APPROVED** [review]
Thanks @rmanaem! Nice regex use! I left a couple of suggestions with clarity in mind, take a look.

I do think that we should emit warnings if the `LOCAL_NB_NODES` contains bad input. If that's a bigger deal to address, please make a separate issue to raise that warning so we don't forget

File: app/api/crud.py (hunk 1 above)
Comment (rmanaem) [reply app/api/crud.py:103]: This is a workaround to essentially remove duplicate keys or rather overwrite them.

File: app/api/crud.py (hunk 2 above)
Comment (rmanaem) [reply app/api/crud.py:64]:
```
for node_url, node_name in nodes_dict.items():
        response = util.send_get_request(node_url + "query/", params)

        for result in response:
            result["node_name"] = node_name

        cross_node_results += response

    return cross_node_results
```

Is this what u had in mind?

File: app/api/crud.py (hunk 1 above)
Comment (surchs) [reply app/api/crud.py:103]: I get that, but why do you have to have `TermURL` be both in the key and then again in the value?

**Comment by rmanaem** [comment]
@surchs Made an issue for the warning. See #31

File: app/api/crud.py (hunk 1 above)
Comment (rmanaem) [reply app/api/crud.py:103]: I see. it's supposed to help structure the output at the end to resemble what we have in the node API.

**Comment by surchs** [comment]
:rocket: PR was released in [`v0.1.0`](https://github.com/neurobagel/federation-api/releases/tag/v0.1.0) :rocket:

**Pull request merged**